Changes to Synchrotron Radiation from issue #4921. Part 1.

FilipO28555 opened the pull request:
in validate():
- [x] subtract_functions() is now without a bug and unnecessary list truncation
- [x] Changed the number of bins to 200 for better accuracy

Todos from [issue #4912](https://github.com/ComputationalRadiationPhysics/picongpu/issues/4912):
- [x] use `pathlib` instead of `os.path` [here](https://github.com/ComputationalRadiationPhysics/picongpu/pull/4827#discussion_r1609863570) and [here](https://github.com/ComputationalRadiationPhysics/picongpu/pull/4827#discussion_r1609864862)
- [x] [question on usage](https://github.com/ComputationalRadiationPhysics/picongpu/pull/4827#discussion_r1609879390) of `change_parameters_randomly.py`
- [x] use [`main` function](https://github.com/ComputationalRadiationPhysics/picongpu/pull/4827#discussion_r1609881468) instead of a script like code
- [x] [question on comment](https://github.com/ComputationalRadiationPhysics/picongpu/pull/4827#discussion_r1609882515) regarding output 
- [x] use of `np.asarray` [here](https://github.com/ComputationalRadiationPhysics/picongpu/pull/4827#discussion_r1609891973) and [here](https://github.com/ComputationalRadiationPhysics/picongpu/pull/4827#discussion_r1609899282)
- [x] [question on parameters](https://github.com/ComputationalRadiationPhysics/picongpu/pull/4827#discussion_r1609902008) in `validate.py`
- [x] use seperate functions in python [here](https://github.com/ComputationalRadiationPhysics/picongpu/pull/4827#discussion_r1609908954) and [here](https://github.com/ComputationalRadiationPhysics/picongpu/pull/4827#discussion_r1609909367)
- [x] better way to handle [input arguments](https://github.com/ComputationalRadiationPhysics/picongpu/pull/4827#discussion_r1609940439)

FilipO28555 committed:
change_parameters_randomly.py now has a main()
as well as uses pathlib and some comment corrections

FilipO28555 committed:
1) subtract_functions() is now without a bug and unnecessary list truncation
2) Changed the number of bins to 200 for better accuracy
3) added some printouts for clarity

FilipO28555 committed:
Abstract OpenPMD data reading for photons and electrons in validate.py

- Introduced `read_photon_data` function to encapsulate the logic for reading and processing photon data from OpenPMD series iteration.
- Introduced `read_electron_data` function to encapsulate the logic for reading and processing electron data from OpenPMD series iteration.
- Refactored the `main` function to use the new abstraction functions for reading photon and electron data.

This change improves code readability and maintainability by separating concerns and making the main function less complex.

FilipO28555 committed:
added versions to python requirements

chillenzer commented on share/picongpu/tests/Synchrotron_plugin/lib/change_parameters_randomly.py:
```diff
@@ -23,6 +24,7 @@ def change_param_file(file_name, param_name, value):
                 f.write(line)
 
 
+# change the three parameters
```
Remove. That is redundant.

FilipO28555 replied:
done

chillenzer commented on share/picongpu/tests/Synchrotron_plugin/lib/requirements.txt:
```diff
@@ -1,4 +1,5 @@
-openpmd_api
+numpy==1.26.4
+scipy==1.11.4
+pathlib
```
`pathlib` is part of the standard library, so we don't need to include it.

FilipO28555 replied:
Done

FilipO28555 replied:
The same with os, right?

chillenzer replied:
Yes

chillenzer commented on share/picongpu/tests/Synchrotron_plugin/lib/validate.py:
```diff
@@ -7,15 +7,8 @@
-    y0 = [y for x, y in zip(x0, y0) if x >= minX and x <= maxX]
-    y1 = [y for x, y in zip(x1, y1) if x >= minX and x <= maxX]
-
     y1 = np.interp(x0, x1, y1)  # interpolate y1 to the x0 values
```
Okay, so you are saying that you are fine with the default behaviour of `np.interp` here? I would have assumed that for histograms using `left=0, right=0` would have been the most faithful extrapolation.

FilipO28555 replied:
You are right. I did that. I also changed the name to be more descriptive

chillenzer commented on share/picongpu/tests/Synchrotron_plugin/lib/validate.py:
```diff
@@ -38,57 +31,84 @@ def read_setup(folderName):
+    hist_data = np.abs(p_y * const.c / const.elementary_charge)
+
     if len(hist_data) < 5e5:
         raise SystemExit(
```
Just noticed: Do we really want to raise `SystemExit` here? Raising something a little milder like `ValueError` or a custom `Exception` would have the same effect for running this script. But in case anyone ever tries to build something on top of this, a `ValueError` could be caught and handled appropriately.

FilipO28555 replied:
I changed it to ValueError

chillenzer commented on share/picongpu/tests/Synchrotron_plugin/lib/validate.py:
```diff
@@ -38,57 +31,84 @@ def read_setup(folderName):
     e_w *= e_w_SI  # convert to SI units
+
+    # Reading electron momentum
     h1 = e["momentum"]["y"]
```
Just to be extra sure: The `"y"` in this line has nothing to do with the `"y"` that approximates a $\gamma$ and stands for photons further up, right? It's probably the direction of the momentum?

FilipO28555 replied:
yes, you are right

FilipO28555 replied:
Yes

chillenzer commented on share/picongpu/tests/Synchrotron_plugin/lib/validate.py:
```diff
@@ -114,11 +134,12 @@ def main(dataPath):
 
     x, y = subtract_functions(delta, analytical_integrated, b, a)
-    poorness = np.sum(y) / len(y)  # average error
+    poorness = np.sum(y) / len(y)  # average relative error
```
Use `np.mean` (potentially with `axis=0`) here.

FilipO28555 replied:
We left it because in the future we might do something like np.sum(y*np.sqrt(a)) / len(y)

chillenzer replied:
That's still a mean isn't it? `np.mean(y*np.sqrt(a))`

PrometheusPi replied:
I agree with @chillenzer

PrometheusPi replied:
and why is this now "relative" ?

FilipO28555 replied:
It was always relative.
Also changed to np.mean()

chillenzer reviewed (commented):
Hi, looks good. Thanks for going the extra mile to make this a little more sustainable code. In what sense are the questions answered? They are already ticked but that information has not propagated to me. So, it is likely that whatever method you used to answer them, does not help the uninvolved future reader of your code.

FilipO28555 committed:
Added changes from comments:
 - in np.interp() added left=0 and right=0
 - subtract_functions() is now relative_error_array()
 - SystemExit changed to ValueError

chillenzer reviewed (approved):
Good work!

PrometheusPi commented on share/picongpu/tests/Synchrotron_plugin/lib/validate.py:
```diff
@@ -113,17 +133,18 @@ def main(dataPath):
 if __name__ == "__main__":
     if len(sys.argv[1:]) != 1:
-        raise SystemExit(f"Usage: {sys.argv[0]} <path_to_simulation_data>")
+        raise ValueError(f"Usage: {sys.argv[0]} <path_to_simulation_data>")
```
Just a question: why this change?

chillenzer replied:
I guess that was in response to [this comment](https://github.com/ComputationalRadiationPhysics/picongpu/pull/4926#discussion_r1629120358) applying to a different exception in the same script. The change you are referring to will most likely not make any difference at all and one could argue that not doing it might convey the intent a tiny bit better. So, weakly agreeing with you but probably doesn't matter.

PrometheusPi reviewed (changes requested):
Just some minor requests.

FilipO28555 committed:
Changed to np.mean as suggested
and reverted the ValueError to SystemExit
